README.md
# TZOffset

[![Gem Version](https://badge.fury.io/rb/tz_offset.svg)](http://badge.fury.io/rb/tz_offset)
[![Dependency Status](https://gemnasium.com/molybdenum-99/tz_offset.svg)](https://gemnasium.com/molybdenum-99/tz_offset)
[![Build Status](https://travis-ci.org/molybdenum-99/tz_offset.svg?branch=master)](https://travis-ci.org/molybdenum-99/tz_offset)

Ever tried to convert your distant friend's or colleague's phrase "OK, let's connect tomorrow
at 6 pm my time (GMT+8)" into easily calculatable Ruby code? E.g. what time it would be in your
timezone at 18:00 GMT+8? What is your favorite solution? Now it should be that:

```ruby
TZOffset.parse('GMT+8').local(2016, 10, 20, 18).localtime
# => 2016-10-20 13:00:00 +0300

# or just
TZOffset.parse('+8').local(2016, 10, 20, 18).localtime
# => 2016-10-20 13:00:00 +0300

# Also works with most common timezone abbreviations
TZOffset.parse('CEST').local(2016, 10, 20, 18).localtime
# => 2016-10-20 19:00:00 +0300
```

In other words, `TZOffset` is simple, no-magic, incapsulated abstraction of "time offset".

## Features and problems

* Easy-to-use, intuitive, dead simple, OS independent;
* No brains included: no huge and comprehensive database of historical times, no automatic
  DST conversion; you just know offset you need, and have it as a near-to-mathematical
  value;
* Simple value objects, easily converted to/from YAML (so you can save them to databases,
  pass to delayed jobs and so on);
* Knows about all common timezone abbreviations (got them from
  [Wikipedia list](https://en.wikipedia.org/wiki/List_of_time_zone_abbreviations));
* For ambiguous abbreviations, just returns list of all of them:

```ruby
TZOffset.parse('EET')
# => #<TZOffset +02:00 (EET)>
TZOffset.parse('CDT')
# => [#<TZOffset -05:00 (CDT)>, #<TZOffset -04:00 (CDT)>]
```

* For symbolic timezones, provides a description, if available:

```ruby
TZOffset.parse('CDT').map(&:description)
# => ["Central Daylight Time (North America)", "Cuba Daylight Time"]
TZOffset.parse('CDT').map(&:region)
# => ["Central", "Cuba"]

# NB: Just "Central", "Eastern" and so on is related to North America in timezones nomenclature
```

* for [DST](https://en.wikipedia.org/wiki/Daylight_saving_time) and non-DST timezones provides
  dst-flag and counterpart timezone, if available:

```ruby
eet = TZOffset.parse('EET')
# => #<TZOffset +02:00 (EET)>
eet.dst?
# => false
eet.opposite
# => #<TZOffset +03:00 (EEST)>
[eet.description, eet.opposite.description]
# => ["Eastern European Time", "Eastern European Summer Time"]
```

## Installation

Do your usual routine with gem named `tz_offset` (e.g. `gem install tz_offset` or add
`gem "tz_offset` to your `Gemfile`).

## Usage

Most of it is already shown above!

```ruby
require 'tz_offset'

off = TZOffset.parse('-02:30')
# => #<TZOffset -02:30>
off.now
# => 2016-10-25 16:07:55 -0230
off.local(2016, 10, 1)
# => 2016-10-01 00:00:00 -0230

eet = TZOffset.parse('EET')
# => #<TZOffset +02:00 (EET)>
eet.now
# => 2016-10-25 20:29:03 +0200
eet.description
# => "Eastern European Time"
eet.region
# => "Eastern European"
eet.opposite
# => #<TZOffset +03:00 (EEST)>
eet.opposite.now
# => 2016-10-25 21:39:26 +0300

# Parsing time string into desired timezone
off = TZOffset.parse('-02:30')
off.parse('2014-10-01 12:30')
# => 2014-10-01 12:30:00 -0230
off.parse('12:30')
# => 2016-10-26 12:30:00 -0230
```

## Author

[Victor Shepelev](http://zverok.github.io/) -- extracted from [reality](https://github.com/molybdenum-99/reality)
project.

## License

MIT.


CHANGELOG.md
# TZOffset Changelog

## 0.0.4 - 2018-01-27

* Correct parsing of ±00:00UTC;
* Updated list of known offsets from Wikipedia.

## 0.0.3 - 2017-01-14

* TZOffset is now detailed up to seconds;
* Add simple arithmetics (`+`, unary and binary `-`);
* Add some handy synonims (like `TZOffset.utc` and `TZOffset#utc?`).

## 0.0.2 - 2016-10-25

* Add `TZOffset#parse`.


spec/spec_helper.rb
require 'simplecov'
SimpleCov.start

require 'rspec/its'
require 'timecop'

require 'saharspec'

$:.unshift 'lib'
require 'tz_offset'


spec/tz_offset/tz_offset_spec.rb
describe TZOffset do
  describe '.parse' do
    def from(str)
      described_class.parse(str)
    end

    def minutes_from(str)
      from(str).minutes
    end

    def seconds_from(str)
      from(str).seconds
    end

    it 'can be created from offset str' do
      expect(minutes_from('-01:00')).to eq -60
      expect(minutes_from('+01:00')).to eq 60
      expect(minutes_from('−01:00')).to eq -60
      expect(minutes_from('−02:30')).to eq -150
      expect(minutes_from('+05:45')).to eq 60*5 + 45
      expect(minutes_from('UTC')).to eq 0
      expect(minutes_from('UTC+01:00')).to eq 60
      expect(minutes_from('UTC+2')).to eq 120
      expect(minutes_from('+0200')).to eq 120
      expect(minutes_from('UTC+5:45')).to eq 60*5 + 45
      expect(minutes_from('±00:00UTC')).to eq 0

      expect(seconds_from('5:53:03')).to eq 5*3600 + 53*60 + 3
      expect(seconds_from('-0:01:15')).to eq -(1*60 + 15)
    end

    it 'preserves offset name for named offsets' do
      expect(from('CEST').name).to eq 'CEST'
      expect(from('+0200').name).to be_nil
    end

    it 'parses even offset names unknown to Ruby' do
      expect(minutes_from('CEST')).to eq 120
    end

    it 'returns several results for ambigous tz abbrs' do
      expect(from('CDT'))
        .to be_an(Array)
        .and have_attributes(size: 2)
        .and all(be_a(described_class))
    end

    it 'returns nil for unparseable' do
      expect(from('WTF')).to be_nil
    end
  end

  describe '.zero' do
    specify { expect(described_class.zero).to eq TZOffset.new(0) }
  end

  def off(*val)
    described_class.new(*val)
  end

  describe '#initialize' do
    it 'can be created from number of seconds' do
      expect(off(-3600).seconds).to eq -3600
      expect(off( 3600).seconds).to eq 3600
    end

    it 'can have optional name' do
      expect(off(-3600, name: 'EET').name).to eq 'EET'
      expect(off(-3600).name).to be_nil
    end
  end

  describe '#inspect' do
    def at(*val)
      off(*val).inspect
    end

    it 'works' do
      expect(at(-3600)).to eq '#<TZOffset -01:00>'
      expect(at(+3600)).to eq '#<TZOffset +01:00>'
      expect(at(-9000)).to eq '#<TZOffset -02:30>'
      expect(at(5 * 3600 + 45 * 60)).to eq '#<TZOffset +05:45>'
      expect(at(-(1*60 + 15))).to eq '#<TZOffset -00:01:15>'
    end

    it 'preserves timezone name when available' do
      expect(at(2 * 3600, name: 'CEST')).to eq '#<TZOffset +02:00 (CEST)>'
    end

    it 'shows non-zero seconds' do
      expect(at(3600 + 1800 + 5)).to eq '#<TZOffset +01:30:05>'
    end
  end

  describe '#to_s' do
    def at(val)
      off(val).to_s
    end

    it 'works' do
      expect(at(-60*60)).to eq '-01:00'
      expect(at(+60*60)).to eq '+01:00'
      expect(at(-150*60)).to eq '-02:30'
      expect(at(-75)).to eq '-00:01:15'
    end
  end

  describe '#-@' do
    subject { -TZOffset.parse('+01:00') }

    it { is_expected.to eq TZOffset.parse('-01:00') }
  end

  describe '#+' do
    subject { TZOffset.parse('+01:00') + other }

    context 'with other offset' do
      let(:other) { TZOffset.parse('-03:15') }
      it { is_expected.to eq TZOffset.parse('-02:15') }
    end

    context 'with numeric' do
      let(:other) { 120 }
      it { is_expected.to eq TZOffset.parse('+01:02') }
    end

    context 'with anything else' do
      let(:other) { 'xxx' }
      its_block { is_expected.to raise_error ArgumentError }
    end
  end

  describe '#-' do
    subject { TZOffset.parse('+01:00').method(:-) }

    its_call(TZOffset.parse('-03:15')) { is_expected.to ret TZOffset.parse('+04:15') }
    its_call(120) { is_expected.to ret TZOffset.parse('+00:58') }
    its_call('xxx') { is_expected.to raise_error ArgumentError }
  end

  describe '#==' do
    it 'looks at minutes' do
      expect(off(60)).to eq off(60)
      expect(off(60)).to eq off(60, name: 'EET')
      expect(off(60)).not_to eq off(120)
    end

    it 'does not fails on incompatible objects' do
      expect(off(60)).not_to eq 60
    end
  end

  context '<=>' do
    it 'looks at minutes' do
      expect(off(60) <=> off(60)).to eq 0
      expect(off(60) <=> off(60, name: 'EET')).to eq 0
      expect(off(60) <=> off(120)).to eq -1
      expect(off(60) <=> off(30)).to eq 1
    end

    it 'returns nothing on incompatible objects' do
      expect(off(60) <=> 60).to be_nil
    end
  end

  describe '#local' do
    let(:offset) { TZOffset.parse('+5:45') }

    subject { offset.local(2016, 1, 29, 18, 15, 0) }

    it { is_expected.to eq Time.new(2016, 1, 29, 18, 15, 0, '+05:45') }
  end

  describe '#convert' do
    let(:tm) { Time.new(2016, 1, 29, 14, 30, 0, '+02:00') }
    let(:offset) { TZOffset.parse('UTC+5:45') }

    subject { offset.convert(tm) }

    it { is_expected.to eq Time.new(2016, 1, 29, 18, 15, 0, '+05:45') }
    its(:utc_offset) { is_expected.to eq 5*3600 + 45*60 }

    context 'with seconds' do
      let(:offset) { TZOffset.parse('+5:45:20') }

      it { is_expected.to eq Time.new(2016, 1, 29, 18, 15, 20, '+05:45:20') }
      its(:utc_offset) { is_expected.to eq 5*3600 + 45*60 + 20 }
    end
  end

  describe '#now' do
    let(:tm) { Time.new(2016, 1, 29, 14, 30, 0, '+02:00') }

    before{
      Timecop.freeze(tm)
    }

    let(:offset) { TZOffset.parse('+5:45') }
    subject { offset.now }

    it { is_expected.to eq Time.new(2016, 1, 29, 18, 15, 0, '+05:45') }
    its(:utc_offset) { is_expected.to eq 5*3600 + 45*60 }
  end

  describe '#parse' do
    let(:tm) { Time.new(2016, 1, 29, 14, 30, 0, '+02:00') }
    let(:offset) { TZOffset.parse('+5:45') }

    before{
      Timecop.freeze(tm)
    }

    subject { offset.parse('18:15') }

    it { is_expected.to eq Time.new(2016, 1, 29, 18, 15, 0, '+05:45') }
    its(:utc_offset) { is_expected.to eq 5*3600 + 45*60 }
  end

  describe '#zero?' do
    context 'when zero' do
      subject { TZOffset.new(0) }
      it { is_expected.to be_zero }
    end

    context 'when non-zero' do
      subject { TZOffset.new(1) }
      it { is_expected.not_to be_zero }
    end
  end

  describe '#opposite' do
    specify { expect(TZOffset.parse('EET').opposite).to eq TZOffset.parse('EEST') }
    specify { expect(TZOffset.parse('EEST').opposite).to eq TZOffset.parse('EET') }
  end
end


lib/tz_offset/tasks/extract_offsets.rb
class TZOffset
  # @private
  module Tasks
    class ExtractOffsets

      TZ = Struct.new(:abbr, :title, :region, :dst, :val)

      NO_TIME_SUFFIX = [
        'Gambier Islands',
        'South Georgia and the South Sandwich Islands',
        "Heure Avancée d'Europe Centrale francised name for CEST",
        "Heure Avancée d'Europe Centrale French-language name for CEST",
        'Marshall Islands',
        'Indian/Kerguelen'
      ]

      def run
        require 'infoboxer'

        data = Infoboxer.wp.get('List of time zone abbreviations').tables.first.rows[1..-1]
          .map { |r| r.cells.map(&:text_) }
          .reject { |cells| cells[0] == 'UTC' }
          .reject { |cells| cells[1].include?(' (Australia)') } # they are duplicated, like Australian Central Daylight Savings Time vs Central Summer Time (Australia)
          .reject { |cells| cells[2].include?(' - ') } # ASEAN time is range, not exact value
          .map { |abbr, title, val|
            [
              abbr,
              title
                .sub(/ Same zone as \S+$/, '')
                .sub(/^(#{NO_TIME_SUFFIX.join('|')})$/, '\1 Time'),
              val.sub('±00', '')
            ]
          }
          .map { |abbr, title, val| TZ.new(abbr, title, *parse_title(title), TZOffset.parse(val).minutes) }

        require 'yaml'
        require 'fileutils'

        FileUtils.mkdir_p 'lib/tz_offset/data'

        File.write 'lib/tz_offset/data/abbreviations.yaml', data.sort_by(&:abbr).map(&:to_h).to_yaml
      end

      private

      def parse_title(title)
        title =~ /^(.+?)((?: )(?:Standard|Summer|Daylight|Daylight Savings))? Time(?: \(.+\))?$/i or
          fail(ArgumentError, "Unparseable time title: #{title}")

        region, spec = $1, $2
        [region, /Summer|Daylight/ === spec]
      end
    end
  end
end


lib/tz_offset.rb
require 'time'

# Simple class representing timezone offset (in minutes). Knows almost nothing
# about timezone name, DST or other complications, but useful when ONLY offset is known.
#
# Usage:
#
# ```ruby
# o = Reality::TZOffset.parse('UTC+3')
# # => #<Reality::TZOffset(UTC+03:00)>
#
# o.now
# # => 2016-04-16 19:01:40 +0300
# o.local(2016, 4, 1, 20, 30)
# # => 2016-04-01 20:30:00 +0300
# o.convert(Time.now)
# # => 2016-04-16 19:02:22 +0300
# ```
#
class TZOffset
  # Number of seconds in offset.
  #
  # @return [Fixnum]
  attr_reader :seconds

  # Symbolic offset name if available (like "EEST").
  #
  # @note
  #   TZOffset never tries to "guess" the name, it is only known if an object was parsed/created
  #   from it.
  #
  # @return [String]
  attr_reader :name

  # Full symbolic timezone description, as per wikipedia (like "Eastern European Summer Time"
  # for "EEST").
  #
  # @return [String]
  attr_reader :description

  # "Region" part of {#description}, like "Eastern European" for "EEST".
  #
  # @return [String]
  attr_reader :region

  class << self
    # @private
    # ± should hopefully do no harm to be considered minus, as it happens only in ±00:00 statements
    MINUSES = /[−—–±]/

    # Parses TZOffset from string. Understands several options like:
    #
    # * `GMT` (not all TZ names, just well-known
    #   [abbreviations](https://en.wikipedia.org/wiki/List_of_time_zone_abbreviations));
    # * `UTC+3` (or `GMT+3`);
    # * `+03:30`;
    # * ..and several combinations.
    #
    # @example
    #   TZOffset.parse('+05:30') # => #<TZOffset +05:30>
    #   TZOffset.parse('UTC+0200') # => #<TZOffset +02:00>
    #   TZOffset.parse('something') # => nil
    #   TZOffset.parse('EST') # => #<TZOffset -05:00 (EST)>
    #   TZOffset.parse('CST') # => [#<TZOffset -06:00 (CST)>, #<TZOffset +08:00 (CST)>, #<TZOffset -05:00 (CST)>]
    #   TZOffset.parse('CST').map(&:description) # => ["Central Standard Time (North America)", "China Standard Time", "Cuba Standard Time"]
    #
    # @return [TZOffset, Array<TZOffset>, nil] Returns array on ambigous abbreviations, and `nil` on
    #   unparseable string.
    def parse(text)
      return ABBREV[text.upcase] if ABBREV.key?(text.upcase)

      sec = parse_text(text.gsub(MINUSES, '-'))
      sec && new(sec)
    end

    # Returns zero (UTC) offset.
    #
    # @return [TZOffset]
    def zero
      @zero ||= new(0)
    end

    alias_method :utc, :zero

    private

    def parse_text(text)
      case text
      when /^[A-Z]{3}$/
        Time.zone_offset(text)
      when /^(?:UTC|GMT)?([+-]\d{1,2}:?\d{2})$/
        parse_zone_offset($1)

      when /^(?<sign>[+-]?)(?<hours>\d{1,2})(:(?<minutes>\d{2})(:(?<seconds>\d{2}))?)?$/
        parse_with_seconds(Regexp.last_match)

      when /^(?:UTC|GMT)?([+-]\d{1,2})/
        $1.to_i * 3600
      end
    end

    def parse_zone_offset(offset)
      Time.zone_offset(offset.sub(/^([+-])(\d):/, '\10\2:'))
    end

    def parse_with_seconds(match)
      (match[:sign] == '-' ? -1 : +1) *
        (match[:hours].to_i * 3600 + match[:minutes].to_i * 60 + match[:seconds].to_i)
    end
  end

  # Constructs offset from number of seconds. In most cases, you don't
  # want to use it, but rather {TZOffset.parse}.
  #
  # @param seconds [Fixnum] Number of seconds in offset.
  def initialize(seconds, name: nil, description: nil, region: nil, isdst: nil)
    @seconds = seconds
    @name = name
    @description = description
    @isdst = isdst
    @region = region
  end

  # Minutes component of offset.
  #
  # @return [Integer]
  def minutes
    seconds.abs / 60 * (seconds <=> 0)
  end

  alias_method :to_i, :seconds

  # @return [String]
  def inspect
    nm = name ? " (#{name})" : ''

    '#<%s %s%02i:%02i%s%s>' %
      [self.class.name, sign, *minutes.abs.divmod(60), inspectable_seconds, nm]
  end

  # @return [String]
  def to_s
    '%s%02i:%02i%s' % [sign, *minutes.abs.divmod(60), inspectable_seconds]
  end

  # @return [TZOffset] Offset negated.
  def -@
    TZOffset.new(-seconds)
  end

  # Sums offset with other or just number of seconds.
  #
  # @example
  #   TZOffset.parse('+2') + TZOffset.parse('+5')
  #   # => #<TZOffset +07:00>
  #   TZOffset.parse('+2') + 1200
  #   # => #<TZOffset +02:20>
  #
  # @return [TZOffset]
  def +(other)
    case other
    when TZOffset
      TZOffset.new(seconds + other.seconds)
    when Numeric
      TZOffset.new(seconds + other)
    else
      fail ArgumentError, "Can't sum with #{other.class}"
    end
  end

  # Substracts other offset or number of seconds.
  #
  # @example
  #   TZOffset.parse('+2') - TZOffset.parse('+5')
  #   # => #<TZOffset -03:00>
  #   TZOffset.parse('+2') - 1200
  #   # => #<TZOffset +01:40>
  #
  # @return [TZOffset]
  def -(other)
    other.respond_to?(:-@) or fail ArgumentError, "Can't subtract #{other.class} from TZOffset"
    self + -other
  end

  # If offset is symbolic (e.g., "EET", not just "+02:00"), returns whether it is daylight
  # saving time offset.
  #
  # See also {#opposite} for obtaining non-DST related to current DST offset and vice versa.
  def dst?
    @isdst
  end

  # If this offset is zero (UTC).
  def zero?
    @seconds.zero?
  end

  alias_method :utc?, :zero?

  # @return [Boolean]
  def <=>(other)
    return nil unless other.is_a?(TZOffset)
    minutes <=> other.minutes
  end

  include Comparable

  # @return [Boolean]
  def ==(other)
    other.class == self.class && other.seconds == seconds
  end

  # Like Ruby's `Time.local`, but in current offset.
  #
  # @return [Time] Constructed time in that offset.
  def local(*values)
    values << 0 until values.count == 6
    mk(*values)
  end

  # Converts `tm` into current offset.
  #
  # @param tm [Time] Time object to convert (with any offset);
  # @return [Time] Converted object.
  def convert(tm)
    t = tm.getutc + seconds

    # FIXME: usec are lost
    mk(t.year, t.month, t.day, t.hour, t.min, t.sec)
  end

  # Like Ruby's `Time.now`, but in current offset.
  #
  # @return [Time] Current time in that offset.
  def now
    convert(Time.now)
  end

  # For symbolic offsets, returns opposite (DST for non-DST and vice versa) offset object if
  # known, `nil` otherwise.
  #
  # @example
  #   eet = TZOffset.parse('EET')
  #   # => #<TZOffset +02:00 (EET)>
  #   eet.opposite
  #   # => #<TZOffset +03:00 (EEST)>
  #   TZOffset.parse('+8').opposite
  #   # => nil
  #
  # @return [TZOffset]
  def opposite
    return nil unless region
    ABBREV.values.flatten.detect { |tz| tz.region == region && tz.dst? == !dst? }
  end

  # Like `Time.parse`, but produces time in current offset.
  #
  # @note
  #   If time string contains timezone abbreviation or offset by itself, it is just ignored.
  #   The method is intended for "quick-n-dirty" parsing of things like "what is 12:30 in
  #   CEST currently?"
  #
  # @example
  #   # My current date and zone
  #   Time.parse('12:30')
  #   # => 2016-10-26 12:30:00 +0300
  #   TZOffset.parse('PDT').parse('12:30')
  #   # => 2016-10-26 12:30:00 -0700
  #   # Now I can know what it will be for me when somebody
  #   # says "I'll be available 12:30 PDT tomorrow":
  #   TZOffset.parse('PDT').parse('12:30').localtime
  #   # => 2016-10-26 22:30:00 +0300
  #
  # @param str [String] Same string that `Time.parse` could accept.
  # @return [Time] Time parsed from `str` into current offset.
  def parse(str)
    t = Time.parse(str)
    t && local(t.year, t.month, t.day, t.hour, t.min, t.sec)
  end

  private

  def inspectable_seconds
    (seconds.abs % 60).zero? ? '' : ':%02i' % (seconds.abs % 60)
  end

  def sign
    minutes < 0 ? '-' : '+'
  end

  def mk(*components)
    Time.new(*components, to_i)
  end
end

require_relative 'tz_offset/abbrev'
